packages/next-swc/crates/napi/src/next_api/project.rs

```diff
@@ -607,3 +612,122 @@ pub fn project_update_info_subscribe(
+    #[napi(ts_arg_type = "{ __napiType: \"Project\" }")] project: External<ProjectInstance>,
+    frame: StackFrame,
+) -> napi::Result<Option<TracedSource>> {
+    let turbo_tasks = project.turbo_tasks.clone();
+    let traced_frame = turbo_tasks
+        .run_once(async move {
+            let file = match Url::parse(&frame.file) {
+                Ok(url) => match url.scheme() {
+                    "file" => url.path().to_string(),
+                    _ => bail!("Unknown url scheme"),
+                },
+                Err(_) => frame.file.to_string(),
+            };
+
+            let Some(chunk_base) = file.strip_prefix(
+                &(format!(
+                    "{}/{}",
+                    project.container.project().await?.root_path,
+                    project.container.project().dist_dir().await?
+                )),
+            ) else {
+                // File doesn't exist within the dist dir
+                return Ok(None);
+            };
+
+            let chunk_path = format!(
+                "{}{}",
+                project
+                    .container
+                    .project()
+                    .client_relative_path()
+                    .await?
+                    .path,
+                chunk_base
+            );
+
+            let path = project
+                .container
+                .project()
+                .client_root()
+                .fs()
+                .root()
+                .join(chunk_path);
+
+            let generatable: Vc<Box<dyn GenerateSourceMap>> =
+                Vc::try_resolve_sidecast(project.container.get_versioned_content(path))
+                    .await?
+                    .context("Chunk cannot produce a sourcemap")?;
+
+            let map = generatable
+                .generate_source_map()
+                .await?
+                .context("Chunk is missing a sourcemap")?;
+
+            let token = map
+                .lookup_token(frame.line as usize, frame.column.unwrap_or(0) as usize)
+                .await?
+                .clone_value()
+                .context("Unable to trace token from sourcemap")?;
+
+            let Token::Original(token) = token else {
+                bail!("Expected token to be an OriginalToken")
+            };
+
+            let Some(source_file) = token.original_file.strip_prefix("/turbopack/[project]/")
+            else {
+                bail!("Original file outside project")
+            };
+
+            let source_content = &*project
+                .container
+                .project()
+                .project_path()
+                .join(source_file.to_string())
+                .read()
+                .await?;
+
+            let FileContent::Content(source_content) = source_content else {
+                bail!("No content for source file")
+            };
+
+            let mut source = "".to_string();
+            Read::read_to_string(&mut source_content.read(), &mut source)?;
```

Comment:
We already have this file in memory from the `VersionedContent`. If we derive the `Vc<FileSystemPath>`, we can use the rope reader to get the contents without hitting disk.